Allow custom header and back button to use the props from react-navigation

vonovak opened the pull request:
as outline in [screen navigation options](https://reactnavigation.org/docs/navigators/stack#Screen-Navigation-Options), navigators accept props for rendering custom header and back button.

I may pass a custom react element (`MyHeaderLeft`) as `headerLeft ` and it will get rendered [here](https://github.com/react-community/react-navigation/blob/df788fd9ac22ee93df27f87ec21958234fa97d76/src/views/Header/Header.js#L122). 
However, if I do that, `MyHeaderLeft` will miss out on all the props that the standard `HeaderBackButton` receives from react-navigation [here](https://github.com/react-community/react-navigation/blob/df788fd9ac22ee93df27f87ec21958234fa97d76/src/views/Header/Header.js#L136). I will somehow have to hardcode those things (eg. the styles) into `MyHeaderLeft`, which is not nice.

In my scenario, I want `MyHeaderLeft` to

1 . get all the props from react-navigation
2 . get one (or more) more extra prop that I will provide
3 . based on some condition, `MyHeaderLeft` will render the standard `HeaderBackButton` from react-navigation or something else.

to illustrate the usage:

```js
  import { withProps } from 'recompose';
  ...
  static navigationOptions = ({ navigation }) => ({
   //selection is my custom object
    HeaderLeftComponent: withProps({ selection: navigation.state.params.selection})
     (MyHeaderLeft),
  });
```

MyHeaderLeft
```js
import { HeaderBackButton } from 'react-navigation';

export class MyHeaderLeft extends React.Component {
  render() {
    // I get my custom selection prop here, and also all the others from react-naviagtion!
    const { selection, onPress, ...other } = this.props;  

    // maybe I want to render something else
    if (someCondtition) {
      return null;
    }
    // maybe I want to modify the onPress prop here
    return <HeaderBackButton {...other} onPress={myOnPress} />;
  }
}
```

So far, this PR only adds the functionality to the code. First I want to find out if something like this would be merged, and if so, we can talk about the naming of the props, and then I can write up the docs too. Thanks!

vonovak commented:
I have found 3 PRs that attempt to solve more or less the same problem (#1483 #1538 #1291 ), and imho this PR does so with the greatest amount of flexibility and transparency without breaking current behavior.

kelset commented:
I think this PR is solid :) Just one question, should this change be "explained" in the docs? I feel it should be mentioned.

vonovak commented:
yes, this has to be documented. I noticed @spencercarli assigned himself, so I'll be waiting for his comments. Imo the only thing left to decide is if the prop should be named `HeaderTitleComponent` or if the name should be different.

spencercarli commented on src/views/Header/Header.js:
```diff
@@ -149,8 +150,9 @@ class Header extends React.PureComponent<Props, State> {
     const width = this.state.widths[props.scene.key]
       ? (this.props.layout.initWidth - this.state.widths[props.scene.key]) / 2
       : undefined;
+    const UsedBackButton = options.HeaderLeftComponent || HeaderBackButton;
```
What's the difference between `HeaderLeftComponent` and `headerLeft`? I think this API would be confusing - it doesn't seem consistent with other naming conventions used.

vonovak replied:
Thanks for the question @spencercarli! The difference is that `headerLeft` is an element, but `HeaderLeftComponent` is a component. The motivation for the change is described in the description, and the naming was inspired by react-native core, for example by [FlatList](https://github.com/facebook/react-native/blob/515eb0e8012a7a8f085a8e410c6c694011fd8c1d/Libraries/Lists/FlatList.js#L94). BUT looking at the code I see that `ListHeaderComponent?: ?(React.ComponentType<any> | React.Element<any>),`, ie. `ListHeaderComponent` can be an element or a component. 

So, let's not introduce another prop but take advantage of the existing `headerLeft`: my suggestion is that `headerLeft`, which now can be a react element, can also be a react component. Based on whether it is an element or component, we will render the appropriate thing: if it is an element, we just return it, and if it is a component, we create an element out of it, and return that. Just like [in VirtualizedList](https://github.com/facebook/react-native/blob/515eb0e8012a7a8f085a8e410c6c694011fd8c1d/Libraries/Lists/VirtualizedList.js#L752).

spencercarli replied:
I think that's a great idea, @vonovak!

spencercarli commented on src/views/Header/Header.js:
```diff
@@ -123,14 +123,15 @@ class Header extends React.PureComponent<Props, State> {
+    const UsedHeaderTitle = details.options.HeaderTitleComponent || HeaderTitle;
     return (
-      <HeaderTitle
+      <UsedHeaderTitle
```
Not a fan of `Used` for naming. Perhaps `RenderedHeaderTitle`?

vonovak replied:
I agree

spencercarli commented:
Sorry for my radio silence. Just left some comments.

vonovak commented:
@spencercarli I have updated the code and added documentation. I had to use `$FlowFixMe` to suppress some flow errors that I couldn't get rid of no matter what. Feel free to try to fix them.
 (Actually, the code in react-native's `VirtualizedList` also does that, although I don't know if it's for the same reason.)

spencercarli commented:
Looks good, @vonovak!
